### ISSUE TITLE
Implement Multi-Factor Authentication

### ISSUE THREAD

**Issue by Datamance** [lookit/lookit-api#511]
**TL;DR**: Implement Multi-factor authentication requirements Experimenters and Staff members only.

**Narrative**
As a staff member for Lookit, I want to mitigate the risk of data breaches and user account hacks, while preserving a relatively frictionless participation experience where possible.

As a researcher using Lookit, I want my experimental data to be as secure as possible - even if a hacker gets my password, they shouldn't be able to acquire the potentially sensitive participant data that I have access to. Multifactor authentication is the gold standard for this. 

As a participant, I would prefer a low barrier to entry in getting started with Lookit.

**Acceptance Criteria**

- [x] Experimenter login has 2FA required
- [x] Automated Tests:
  ### Registration
  - [x] Full correct flow: Researcher registers, is automatically redirected to the 2FA setup page, and verifies with a correct code.
  - [x] Bad first-time verification reloads QR: Researcher registers, but enters wrong OTP during first-time verification, resulting in the QR page reloading.
  ### Login
  - [x] Regular redirects to 2FA verification page: Researcher is redirected to 2FA verification after logging in normally.
  - [x] Regular login can't access `/exp/` views: Regularly logged-in researcher is redirected to 2FA verification page when attempting to visit `/exp/` urls.
  - [x] 2FA login success: Researcher can successfully verify OTP auth code
  - [x] 2FA login fail: Researcher is redirected back to OTP verification page after failed verification
  - [x] Participant login functions correctly
  - [x] Participant doesn't have access to 2FA views
  ### QA Checklist:
- [x] New participant can create and use account
  - [x] Login/signup page is easy to find, little chance of accidentally/out of mild curiosity signing up as researcher
  - [x] Can view flat pages and studies page/study detail pages without login
  - [x] Receive appropriate error if trying to register with email already associated with a researcher or participant account, or registering with something that doesn't look like an email address
  - [x] Receive appropriate error if password does not meet requirements
  - [x] Not prompted for OTP during signup or login
  - [x] Redirected appropriately upon successful signup
  - [x] Can create and edit child, create and edit demographic info, edit email info
  - [x] Can edit password (appropriate error if new password doesn't match or meet requirements)
  - [x] Can log out and log back in
  - [x] Can request password reset while not logged in, and use link to reset password (or log in with existing password)
  - [x] Views that require login appropriately blocked or redirected when not logged in: study participation, account/child/demographic
  - [x] Views that require experimenter login appropriately blocked or redirected when logged in as participant
  - [x] Per recent discussion - participant does not have the option to set up 2FA
- [x] New researcher can create and use account
  - [x] Login/signup page is reasonably easy to find; naive researcher looking around can find and is unlikely to accidentally create a participant account
  - [x] Receive appropriate error if trying to register with email already associated with a researcher or participant account, or registering with something that doesn't look like an email address
  - [x] Receive appropriate error if password does not meet requirements
  - [x] Redirected to set up 2FA upon successful signup
  - [x] Can enable 2FA, set up immediately, access exp pages
  - [x] Can enable 2FA, NOT set up immediately, log out and back in (without being prompted for OTP), not access exp pages, set up 2FA
  - [x] Can create and edit child, create and edit demographic info, edit email info
  - [x] Can edit password (appropriate error if new password doesn't match or meet requirements)
  - [x] Can log out and log back in
  - [x] Can request password reset while not logged in, and use link to reset password (or log in with existing password) but any existing 2FA is left intact
  - [x] Views that require login appropriately blocked or redirected when not logged in: study participation, account/child/demographic (can be either blocked when not logged in + 2FA, or only require password authentication)
  - [x] Views that require experimenter login appropriately blocked or redirected when not logged in or when logged in without 2FA, with indication to user that 2FA needs to be enabled:
    - [x] because have only gotten halfway through the login process (have done password authentication but not OTP yet)
    - [x] because 2FA has not been enabled yet/no OTP secret
    - [x] because there is an OTP secret but it has not been set up with successful test yet
    - [x] because 2FA was at one point enabled but has been disabled
  - [x] Exp views are accessible & render successfully if logged in with 2FA (already covered by tests I think)
  - [x] Can disable 2FA
    - [x] then immediately unable to access experimenter views
    - [x] can log out and back in; not prompted for OTP
    - [x] can still view participant views (account, participation)
    - [x] can re-enable 2FA
  - [x] Existing researcher can add 2FA and use account
    - [x] Researcher with existing OSF account cannot log in, but can request password reset, then change password
    - [x] Researcher is then unable to access exp views until activating 2FA
    - [x] Researcher can enable 2FA, enter test code, and view exp views
    - [x] This researcher (i.e. with previously-OSF-created account) can view and edit own children, change password, and exp views render properly
  - [x] Admin can reset OTP secret for a given account using the admin interface. (Probably want to make Users searchable by email address for convenience.) Upon reset, researcher can log in, is not prompted for OTP, and can enable 2FA from account page. 

**Implementation Notes**
~We can enable 2FA one of two ways:~
~1) building 2FA into the native login capabilities (using django-MFA: https://github.com/MicroPyramid/django-mfa)~ 
~or~ 
~2) leveraging django-allauth to enable Google or Facebook authenticated accounts, similar to how we do for OSF experimenter accounts.~

After extensive review of the options and given our unique constraints (only enforcing 2FA for certain views, and Experimenter/Staff accounts only), we decided to implement 2FA using a combination of PyOTP and custom django models/views/auth backends/forms.

**Comment by Datamance** [comment]
### QA comments inline:
> * Login/signup page is easy to find, little chance of accidentally/out of mild curiosity signing up as researcher

The login and signup links are where they've always been. If you'd like to add another link somewhere, or change the location of the current login and signup link let me know and I'll make the change.

>* Redirected appropriately upon successful signup

Currently participant signup redirects to `/account/demographics/` and researcher account creation to `2fa-setup`. This seems in line with what we want but if you feel differently let me know.


> * Login/signup page is reasonably easy to find; naive researcher looking around can find and is unlikely to accidentally create a participant account

The researcher registration form is somewhat occluded, by design:
![image](https://user-images.githubusercontent.com/8699411/88709557-59256d00-d0e3-11ea-806e-cc64bc123a4b.png)

You [OK'd this earlier](https://lookit-mit.slack.com/archives/DDUSDNVLJ/p1595279127007900) on Slack. If you want something different, let me know.

> * can log out and back in; not prompted for OTP

Researchers aren't prompted for a current verification code, but they are redirected to `/2fa-setup/`. 

> * Researcher with existing OSF account cannot log in, but can request password reset, then change password

There's not a great way to test this _directly_ at this point, but
1. As the system stands now, "a user is a user is a user" - there is nothing to suggest that users of any kind won't be able to request a password reset via the existing process, as long as they have an email in the DB.
2. We can always test this in staging to double check.

> * Admin can reset OTP secret for a given account using the admin interface. (Probably want to make Users searchable by email address for convenience.) Upon reset, researcher can log in, is not prompted for OTP, and can enable 2FA from account page.

There is a new "Google authenticator totps" section under the accounts subheading of the admin view. You can search by username (email) and nickname, and delete the OTP of interest as necessary.
